Policies: Adds deprecation policy (#68439)
grafana/grafana: contribute/deprecation-policy.md

## Duration

| Size   | Duration   | Example                                                          |
| ------ | ---------- | ---------------------------------------------------------------- |
| Large  | 1-2 years  | Classic alerting, scripted dashboards, AngularJS                 |
| Medium | 6 months   | Supported Database for Grafana's backend                         |
| Small  | 1-3 months | Refresh OAuth access_token automatically using the refresh_token |

## Announced deprecations.

| Name                  | Annoucement Date | Disabling date | Removal Date | Description                                                                                                               | Status  |
| --------------------- | ---------------- | -------------- | ------------ | ------------------------------------------------------------------------------------------------------------------------- | ------- |

> | Support for Mysql 5.7 | 2023-05-15       | October 2023   |              | MySQL 5.7 is getting deprecated in October 2023 and Grafana's policy is to test against the officially supported version. | Planned |

Author: I'll write up an issue and link to it before merging.

# Deprecation policy

Prior to Grafana 10, major releases were an opportunity to make breaking changes and remove code we didn’t want to maintain.

> Going forward that is something we will avoid in order to make it easier to upgrade Grafana. From now on we should avoid making unexpected breaking changes to the customer experience and our ability to operate our service.

Reviewer: ```suggestion
Going forward we will do our best to **not** make breaking changes, in order to make upgrading Grafana as easy and reliable as possible. From now on we will avoid making unexpected breaking changes to the customer experience and our ability to operate our service.
```

Author: kind of passive voice 🤔 
yours is better but I might take another stab at this one

> Prior to Grafana 10, major releases were an opportunity to make breaking changes and remove code we didn’t want to maintain.
> 
> Going forward we will do our best to **not** make breaking changes, in order to make upgrading Grafana as easy and reliable as possible. From now on we will avoid making unexpected breaking changes to the customer experience and our ability to operate our service.

Reviewer 2: This has always from day 1 been a major goal and focus (to make upgrading as easy and as pain-free as possible). So not sure what is changing (other than adding a formal policy). This seems to conflate unexpected breaking changes and deprecating features. 

I would rephrase this a bit. 

```suggestion
Major releases are when we will be making breaking changes, like changing behavior or removing a feature. 

We always do our best **not** to introduce breaking changes in order to make upgrading Grafana as easy and reliable as possible. However, at times we have to introduce a breaking change by changing behavior or by removing a feature.  

But in order to remove a feature and introduce a breaking change we need to create a deprecation plan. 
```

> If we cannot remove the features without introducing a breaking chance we should follow an deprecation plan as listed below.

Reviewer: ```suggestion
If we cannot remove a feature without introducing a breaking chance we should create a deprecation plan as listed below:
```

> If we cannot remove a feature without introducing a breaking chance we should create a deprecation plan as listed below:

Reviewer: Typo fix and stronger commitment
```suggestion
If we cannot remove a feature without introducing a breaking change we will create a deprecation plan as listed below:
```

> If we cannot remove a feature without introducing a breaking change we will create a deprecation plan as listed below:

Reviewer 2: nitpick: These two sentences seem to say almost the thing and make it seem like a list will follow.

Author: Solved :)

- A deprecation plan should have the following steps.

> - Figure out usage stats for the feature.

Reviewer: ```suggestion
- Determine usage levels of the feature.
```

> - A deprecation plan should have the following steps.

Reviewer 2: This is the third "a deprecation plan that includes" in row :smile:   (see the two sentences above)

- Find alternative solutions and possible migration paths.
- Announce deprecation of the feature.
- Migrate users if possible
- Give users time to adjust to the deprecation.
- Disable the feature by default.
- Remove the feature from the code base.

Depending on the size and importance of the feature this can be a design doc or an issue. We want this to be written communication for all parties so we know it's intentional and that did a reasonable attempt to avoid breaking changes unless needed. The size of the feature also means different notice times between Depreciation and disabling as well as disabling and removal. The actual duration will depend on releases of Grafana and the table below should be used as a guide.

Grafana employees can find more details in our internal docs.

> ## Grace period between announcement and disablement

Reviewer: Not sure disablement is a word
```suggestion
## Grace period between announcement and disabling feature by default
```
